### PR TITLE
PMM-3810 Implement `pmm-admin inventory add node`.

### PR DESCRIPTION
**Pull request by BupycHuk**

### CODE REVIEW

**Comment by it-percona** [comment]
[CLA assistant check](https://cla-assistant.percona.com:443/percona/pmm-admin?pullRequest=2) <br/>All committers have signed the CLA.

File: main.go [hunk 1]
```diff
@@ -21,6 +21,7 @@ import (
 	"crypto/tls"
 	"encoding/json"
 	"fmt"
+	"github.com/percona/pmm-admin/commands/inventory"
```
Comment (AlekSi) [inline main.go]: Move that import below.

**Comment by AlekSi** [comment]
Looking into CI failure…

**Review by AlekSi — CHANGES_REQUESTED** [review]
```
$ pmm-admin --help
pmm-admin: error: duplicate command "node", try --help
```

File: commands/inventory/add_node.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// Package inventory provides inventory commands.
```
Comment (AlekSi) [inline commands/inventory/add_node.go]: Package-level comment should be present only in one file.

File: commands/inventory/add_node_generic.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+Node name: {{ .Node.NodeName }}
+
+MachineID    : {{ .Node.MachineID }}
+Distro       : {{ .Node.Distro }}
+DistroVersion: {{ .Node.DistroVersion }}
+Address      : {{ .Node.Address }}
+`)
+
+type addNodeGenericResult struct {
+	Node *nodes.AddGenericNodeOKBodyGeneric
```
Comment (AlekSi) [inline commands/inventory/add_node_generic.go]: Add `json` tag for `--json` output.

File: commands/inventory/inventory.go [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+package inventory
+
+import (
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+// register command
+var (
+	InventoryC    = kingpin.Command("inventory", "Inventory commands.")
+	InventoryAddC = InventoryC.Command("add", "Add to inventory commands.")
```
Comment (AlekSi) [inline commands/inventory/inventory.go]: Both should not be exported.

File: commands/inventory/add_node_container.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+package inventory
+
+import (
+	"github.com/percona/pmm/api/inventory/json/client"
+	"github.com/percona/pmm/api/inventory/json/client/nodes"
+
+	"github.com/percona/pmm-admin/commands"
+)
+
+var addContainerNodeResultT = commands.ParseTemplate(`
```
Comment (AlekSi) [inline commands/inventory/add_node_container.go]: `addNodeContainerResultT`

File: commands/inventory/add_node_container.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+Container Node added.
+Node ID  : {{ .Node.NodeID }}
+Node name: {{ .Node.NodeName }}
+
+MachineID          : {{ .Node.MachineID }}
+DockerContainerID  : {{ .Node.DockerContainerID }}
+DockerContainerName: {{ .Node.DockerContainerName }}
+`)
+
+type addContainerNodeResult struct {
```
Comment (AlekSi) [inline commands/inventory/add_node_container.go]: `addNodeContainerResult`

File: commands/inventory/add_node_container.go [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+	resp, err := client.Default.Nodes.AddContainerNode(params)
+	if err != nil {
+		return nil, err
+	}
+	return &addContainerNodeResult{
+		Node: resp.Payload.Container,
+	}, nil
+}
+
+// register command
```
Comment (AlekSi) [inline commands/inventory/add_node_container.go:81]: commands

File: commands/inventory/add_node_generic.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+package inventory
+
+import (
+	"github.com/percona/pmm/api/inventory/json/client"
+	"github.com/percona/pmm/api/inventory/json/client/nodes"
+
+	"github.com/percona/pmm-admin/commands"
+)
+
+var addGenericNodeResultT = commands.ParseTemplate(`
```
Comment (AlekSi) [inline commands/inventory/add_node_generic.go]: addNodeGenericResultT

File: commands/inventory/add_node_generic.go [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+	resp, err := client.Default.Nodes.AddGenericNode(params)
+	if err != nil {
+		return nil, err
+	}
+	return &addNodeGenericResult{
+		Node: resp.Payload.Generic,
+	}, nil
+}
+
+// register command
```
Comment (AlekSi) [inline commands/inventory/add_node_generic.go:82]: commands

**Review by AlekSi — CHANGES_REQUESTED** [review]

**Comment by AlekSi** [comment]
Also it looks like commands are not registered:
```
pmm-admin add -h
usage: pmm-admin add <command> [<args> ...]

Add Service to monitoring.

Flags:
  -h, --help                   Show context-sensitive help (also try --help-long and --help-man).
      --version                Show application version.
      --server-url=SERVER-URL  PMM Server URL.
      --server-insecure-tls
      --debug                  Enable debug logging.
      --trace                  Enable trace logging (implies debug).
      --json                   Enable JSON output.

Subcommands:
  add mysql [<flags>] [<address>] [<name>]
    Add MySQL to monitoring.
```

**Comment by BupycHuk** [comment]
try pmm-admin inventory add -h

**Comment by AlekSi** [comment]
My bad. :) `pmm-admin -h` works.

Please also see updated ticket about custom labels.

File: commands/inventory/add_node_container.go (hunk 7 above)
Comment (BupycHuk) [reply commands/inventory/add_node_container.go:81]: why commands?

Comment (AlekSi) [reply commands/inventory/add_node_container.go:81]: My bad. :)

File: commands/inventory/add_node_generic.go (hunk 9 above)
Comment (AlekSi) [reply commands/inventory/add_node_generic.go:82]: wrong

File: commands/inventory/add_node_container.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+func (res *addNodeContainerResult) Result() {}
+
+func (res *addNodeContainerResult) String() string {
+	return commands.RenderTemplate(addNodeContainerResultT, res)
+}
+
+type addNodeContainerCommand struct {
+	NodeType            string
```
Comment (AlekSi) [inline commands/inventory/add_node_container.go]: Not used, remove.

File: commands/inventory/add_node_container.go [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+	NodeName            string
+	MachineID           string
+	DockerContainerID   string
+	DockerContainerName string
+	Address             string
```
Comment (AlekSi) [inline commands/inventory/add_node_container.go:51]: Should be used (for now).

File: commands/inventory/add_node_container.go [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+	CustomLabels        string
+}
+
+func (cmd *addNodeContainerCommand) Run() (commands.Result, error) {
+	customLabels, err := parseCustomLabels(cmd.CustomLabels)
```
Comment (AlekSi) [inline commands/inventory/add_node_container.go:56]: `err` is ignored.

File: commands/inventory/add_node_generic.go [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+func (res *addNodeGenericResult) Result() {}
+
+func (res *addNodeGenericResult) String() string {
+	return commands.RenderTemplate(addNodeGenericResultT, res)
+}
+
+type addNodeGenericCommand struct {
+	NodeType      string
```
Comment (AlekSi) [inline commands/inventory/add_node_generic.go]: Not used, remove.

File: commands/inventory/add_node_container.go [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+	AddNodeContainerC = addNodeC.Command("container", "Add container node to inventory.")
+)
+
+func init() {
+	AddNodeContainerC.Arg("name", "Node name").StringVar(&AddNodeContainer.NodeName)
+
+	AddNodeContainerC.Flag("machine-id", "Linux machine-id.").StringVar(&AddNodeContainer.MachineID)
+	AddNodeContainerC.Flag("container-id", "Docker container identifier. If specified, must be a unique Docker container identifier.").StringVar(&AddNodeContainer.DockerContainerID)
+	AddNodeContainerC.Flag("container-name", "Container name.").StringVar(&AddNodeContainer.DockerContainerName)
+	AddNodeContainerC.Flag("address", "Address.").StringVar(&AddNodeContainer.Address)
```
Comment (AlekSi) [inline commands/inventory/add_node_container.go:94]: Custom labels are not set.

**Review by AlekSi — CHANGES_REQUESTED** [review]
Please fix and re-check this PR yourself with attention.

**Review by AlekSi — APPROVED** [review]

**Pull request merged**